### PR TITLE
Implement np.digitize

### LINKED ISSUE #2595: Implement stratified resampling np.digitize in XLA
Hi JAX team,

I'm interested in implementing efficient stratified resampling in JAX. This is a technique for variance reduction in sequential monte carlo algorithms (see section 2.2.1 in Naesseth, Lindsten, and Schön "Elements of Sequential Monte Carlo").

The algorithm provided looks like this

```python
# w.shape = (N, ) 
# x.shape = (N, )
# assert np.sum(w) = 1
def stratified_resampling(key, w, x) :
  N = w.shape[0]
  u = (np.arange(N) + jax.random.uniform(key, (N, )) / N
  bins = np.cumsum(w)
  return x[np.digitize(u, bins)]
```

although I'm also considering a version that doesn't actually do the shuffling in the last step (saving it for a "decoding" process later on which doesn't particularly need to be fast.

The immediate issue is that `np.digitize` is not implemented in JAX. As it were, I don't actually need the gradient for it, but as it's in the hot path and defeats the tracer it's causing some real pain. Even more ideally, I'd like to have something like the interface to `jax.random.categorical` where I can just use unnormalized logits.

How can I help implement at least tracing through `np.digitize`? Alternatively, is there a clever way of doing stratified resampling without using `np.digitize`?

Thanks

### PR DESCRIPTION
**Pull request by jakevdp**
This turned out to be a fairly light wrapper around `searchsorted`

Fixes #2595 
Issue #70

### CODE REVIEW

File: jax/test_util.py [hunk 1]
```diff
@@ -761,7 +761,7 @@ def assertMultiLineStrippedEqual(self, expected, what):
                               msg="Found\n{}\nExpecting\n{}".format(what, expected))
 
   def _CompileAndCheck(self, fun, args_maker, check_dtypes,
-                       rtol=None, atol=None):
+                       rtol=None, atol=None, allow_compiled_sources=False):
```
Comment (mattjj) [inline jax/test_util.py:764]: Can you say a bit more about what `allow_compiled_sources` means? I don't grok it, and also I don't follow the meaning of the name.

Also, why would we get cache misses in repeated calls to `digitize`? Structured control flow shouldn't prevent cache hits. Sorry if I'm being dense :)

File: jax/numpy/lax_numpy.py [hunk 2]
```diff
@@ -3813,6 +3813,16 @@ def searchsorted(a, v, side='left', sorter=None):
   return _searchsorted(a, v, side)
 
 
+@_wraps(onp.digitize)
+def digitize(x, bins, right=False):
+  side = 'right' if not right else 'left'
+  return lax.cond(
+    len(bins) < 2 or bins[1] > bins[0],
```
Comment (mattjj) [inline jax/numpy/lax_numpy.py:3820]: This is really interesting: usually `or` doesn't work in traced code (e.g. we have to write `jit(lambda x, y: x | y)` instead of `jit(lambda x, y: x or y)`), but here because `len(bins) < 2` is a trace-time constant, it does work!

Comment (jakevdp) [reply jax/numpy/lax_numpy.py:3820]: yay for short-circuit operators!

File: jax/test_util.py (hunk 1 above)
Comment (jakevdp) [reply jax/test_util.py:764]: I may have misunderstood what the compile test does, but I observed that the new function fails this test: https://github.com/google/jax/blob/0ea22b7e191a8cfabefeae18834ced2eec80670b/jax/test_util.py#L779-L784

I saw a similar thing previously when making use of other structured control flow routines. I'm not certain why such routines would prevent cache hits, but they appear to do so.

Comment (mattjj) [reply jax/test_util.py:764]: Ah, I realize why it is: it's because of the lambdas! Those create new function objects; even though the functions are equal in some sense, they aren't the same object each time. And caching of higher-order primitives (not just control flow like `lax.cond` but even `jit`) requires reusing the same function object. (An alternative would be for us to define equality of jaxprs, so that when we trace a fresh function object to an equal jaxpr we know we don't have to recompile. But then we've already re-traced a bit wastefully.) We really need some kind of a warning for this...

[Here's the cache in question](https://github.com/google/jax/blob/master/jax/lax/lax_control_flow.py#L58), by the way. If you pass the same function object to that function with equal in_tree and in_avals then we'll get a cache hit.

For this PR, we could either
1. avoid the `_CompileAndCheck` test,
2. add some logic not to check caching (like you have, but perhaps rename the flag to `check_cache_hits`), or
3. revise the implementation not to use lambdas and instead have helper top-level functions.

I think the latter option might be best, but it might be a little tricky and we could punt on it for a follow-up PR. It could look like this:

```python
from jax import linear_util as lu

def digitize(x, bins, right=False):
  side = 'right' if not right else 'left'
  return lax.cond(
    len(bins) < 2 or bins[1] > bins[0],
    (x, bins), lu.hashable_partial(_digitize_helper, side),
    (x, bins[::-1]), lu.hashable_partial(_digitize_helper_reversed, side))

def _digitize_helper(side, x_bins):
  x, bins = x_bins
  return searchsorted(bins, x, side=side)

def _digitize_helper_reversed(side, x_rbins):
  x, rbins = x_rbins
  return len(rbins) - searchsorted(rbins, x, side=side)
```

(The `lu.hashable_partial` utility is like a stripped-down `functools.partial` except the result is hashable in the right way.)

I haven't tested this so my thinking could be buggy here...

Comment (jakevdp) [reply jax/test_util.py]: Ah, that makes sense. I tried using the ``hashable_partial`` approach, but it seems to be geared towards functions that will be bound to primitives rather than functions that will be called.

I ended up rewriting to use ``where`` instead of ``lax.cond`` to avoid the cache issue.

Comment (mattjj) [reply jax/test_util.py]: Oh right, there's a bit more boilerplate needed to `lu.hashable_partial`. I forgot. It's probably not the right tool for this job. We could make such a tool, but the `np.where` solution sounds fine too.
